ocaml/typing/typecore.ml

```diff
@@ -8365,14 +8365,23 @@ and type_let ?check ?check_strict ?(force_toplevel = false)
   let attrs_list = List.map (fun (attrs, _, _, _) -> attrs) spatl in
   let is_recursive = (rec_flag = Recursive) in
 
+  let var_for_binding (_, _, _, pat) =
+    (* The native code toplevel requires special handling of [let _ = ...],
+       see the special case in [Opttoploop.execute_phrase]. *)
+    match pat.ppat_desc with
+    | Ppat_any when force_toplevel ->
+      newvar (Jkind.Primitive.value ~why:Unnamed_toplevel_expression),
+      Jkind.Sort.value
+    | _ -> new_rep_var ~why:Let_binding ()
+  in
```

Comment:
Won't this not catch `let _ : _ = #4.0`? (I thought it might also not catch `let (_ | _) = #4.0`, but that appears to already be unproblematic... probably because `Ppat_constraint` vanishes during type-checking, but `Ppat_or` does not, and `execute_phrase` runs on the result of type-checking.)

Reply:
Good point.  I'd forgotten that even though we now parse the constraints nicely, we mangle them back to the legacy patterns before this point in `type_let`.

And while or-patterns aren't problematic there are a few other forms that are (e.g., `let M.(_) = #3.14`).  Rather than trying to guess which parsetree patterns will be compiled to `Tpat_any`, I've moved this check to typemod after the pattern is translated to typedtree, so we can just check.  Probably should have just done that in the first place.  Now the two cases (`Pstr_eval` and `Pstr_value`) are handled more uniformly.